protocol/messenger.go

```diff
@@ -3279,6 +3271,46 @@ func (m *Messenger) RetrieveAll() (*MessengerResponse, error) {
 	return m.handleRetrievedMessages(chatWithMessages, true, false)
 }
 
+func (m *Messenger) StartRetrieveMessagesLoop(tick time.Duration, cancel <-chan struct{}) {
+	go func() {
+		ticker := time.NewTicker(tick)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				// We might be shutting down here
+				if m == nil {
+					return
+				}
+				m.ProcessAllMessages()
```

Comment:
This could still fail inside `m.ProcessAllMessages()` if `m` becomes `nil` while `ProcessAllMessages` is being executed. Perhaps it's a good idea to add a waitgroup property to `Messenger`, and replace line 3275 by:
```
m.waitGroup.Add(1)
go func() {
    defer m.waitGroup.Done()
```

and later in Shutdown:

```
// Shutdown takes care of ensuring a clean shutdown of Messenger
func (m *Messenger) Shutdown() (err error) {
	close(m.quit)
	m.cancel()
        m.waitgroup.Wait()   // <---------------------------------------------
```

And that way, this goroutine would be safer to execute, IMO

Reply:
This is an interesting finding!
But I think this is not the only place where we start a goroutine that uses messenger. I can fix this one, but who will take care of the others? 😄

Reply:
@richard-ramos renamed `downloadHistoryArchiveTasksWaitGroup` to `shutdownWaitGroup` and resued it here. It was doing just the thing you proposed 🙂